### PR TITLE
properties

### PR DESCRIPTION
**Pull request by rowanc1**
# Properties

In this version `PropMaps` and indiscriminate `mappings` have been removed. A `Depreciate` module has been provided to help with the upgrade process.

Most details are in #444, also includes a lot of whitespace cleaning.

### Features
- WireMap
- Props
- Depreciate

### Happy changes
- tests are now on images and plotting in the examples
- lots of whitespace and PEP8 updates
- Props can be pickled
- Forward simulations without mappings (e.g. set `simga` directly)

### Changes
- dependency on the `properties` library (`v0.2.1`) http://propertiespy.readthedocs.io/en/latest/
- reference to `curModel` has been replaced with `model`

# Examples

Below are a few examples for the `Wires`, `Props` and `Depreciate`.

## Properties

The `properties` library allows for documentation and change `observer`s on things like the `model` change. It also can be serialized and provides auto documentation on the classes. Below is an example contrasting the changes before and after the `PropMap` removal. All classes have been updated internally, however, because of the brittleness of this class, there are some definitions outside of the `SimPEG` codebase.

#### Old code
```python
class EMPropMap(Maps.PropMap):

    sigma = Maps.Property(
        "Electrical Conductivity",
        defaultInvProp=True, 
        propertyLink=('rho', Maps.ReciprocalMap)
    )

    mu = Maps.Property(
        "Inverse Magnetic Permeability",
        defaultVal=mu_0,
        propertyLink=('mui', Maps.ReciprocalMap)
    )

    rho = Maps.Property(
        "Electrical Resistivity",
        propertyLink=('sigma', Maps.ReciprocalMap)
    )
    
    mui = Maps.Property(
        "Inverse Magnetic Permeability",
        defaultVal=1./mu_0,
        propertyLink=('mu', Maps.ReciprocalMap)
    )
```

#### New code

The new `Props` are bound directly on the `Problem` object and there is no default `mapping`. This means that you **must** specify the mapping that you intend to use explicitly.

```python
class BaseEMProblem(Problem.BaseProblem):

    sigma, sigmaMap, sigmaDeriv = Props.Invertible(
        "Electrical conductivity (S/m)"
    )

    rho, rhoMap, rhoDeriv = Props.Invertible(
        "Electrical resistivity (Ohm m)"
    )

    Props.Reciprocal(sigma, rho)

    mu = Props.PhysicalProperty(
        "Magnetic Permeability (H/m)",
        default=mu_0
    )
    mui = Props.PhysicalProperty(
        "Inverse Magnetic Permeability (m/H)"
    )

    Props.Reciprocal(mu, mui)
```

#### Upgrading

The external upgrade process to the new `Props` syntax should be easy. For example, if `sigma` is the physical property that is used for the inversion model, you have to explicitly set that mapping (`sigmaMap`).

```python
prob = Problem(mesh, mapping=my_mapping)
# becomes
prob = Problem(mesh, sigmaMap=my_mapping)
```

To keep backwards compatibility you can import the `Depreciate` module from SimPEG:

```python
from SimPEG import Depreciate
Depreciate.use_old_mappings()
```

This works for both a single mapping and a `PropMap` style list of maps (without slice syntax).

## Wires and Projections

The `SimPEG.Maps.Wires` allows you to create a number of `SimPEG.Maps.Projection` maps that stack a model and parse it out into the various pieces. These pieces can be referred to by name.

```python
mesh = Mesh.TensorMesh([10, 10, 10])

wires = Maps.Wires(
    ('sigma', mesh.nCz),
    ('mu_casing', 1),
)

model = np.arange(mesh.nCz + 1)
assert isinstance(wires.sigma, Maps.Projection)
assert wires.nP == mesh.nCz + 1
named_model = wires * model
named_model.sigma == model[:mesh.nCz]
assert named_model.mu_casing == 10
```

### CODE REVIEW

File: SimPEG/Directives.py [hunk 1]
```diff
@@ -396,38 +414,39 @@ def endIter(self):
...
+    onlyOnStart = False
 
     def initialize(self):
 
         if getattr(self.opt, 'approxHinv', None) is None:
             # Update the pre-conditioner
-            diagA = np.sum(self.prob.G**2.,axis=0) + self.invProb.beta*(self.reg.W.T*self.reg.W).diagonal() #* (self.reg.mapping * np.ones(self.reg.curModel.size))**2.
-            PC     = Utils.sdiag((self.prob.mapping.deriv(None).T *diagA)**-1.)
+            diagA = np.sum(self.prob.G**2., axis=0) + self.invProb.beta*(self.reg.W.T*self.reg.W).diagonal() # * (self.reg.mapping * np.ones(self.reg.curModel.size))**2.
+            PC     = Utils.sdiag((self.prob.mapping.deriv(None).T * diagA)**-1.)
```
Comment (rowanc1) [inline SimPEG/Directives.py]: just a single space.

File: SimPEG/Directives.py [hunk 2]
```diff
@@ -441,9 +460,9 @@ def JtJv(v):
 
                 Jv = self.prob.Jvec(m, v)
 
-                return self.prob.Jtvec(m,Jv)
+                return self.prob.Jtvec(m, Jv)
 
-            JtJdiag = Utils.diagEst(JtJv,len(m),k=self.k)
+            JtJdiag = Utils.diagEst(JtJv, len(m), k=self.k)
```
Comment (lheagy) [inline SimPEG/Directives.py]: Thanks for all of this cleanup!

File: SimPEG/EM/Base.py [hunk 3]
```diff
@@ -1,29 +1,69 @@
...
+    )
+
+    rho, rhoMap, rhoDeriv = Props.Invertible(
+        "Electrical resistivity (Ohm m)"
+    )
+
+    Props.Reciprocal(sigma, rho)
 
+    mu = Props.PhysicalProperty(
+        "Magnetic Permeability",
```
Comment (lheagy) [inline SimPEG/EM/Base.py]: units? (H/m)

File: SimPEG/EM/Base.py [hunk 4]
```diff
@@ -1,29 +1,69 @@
...
-    surveyPair = Survey.BaseSurvey #: The survey to pair with.
-    dataPair = Survey.Data #: The data to pair with.
+    Props.Reciprocal(mu, mui)
 
-    PropMap = EMPropMap #: The property mapping
+    surveyPair = Survey.BaseSurvey  #: The survey to pair with.
+    dataPair = Survey.Data  #: The data to pair with.
+
+    # PropMap = EMPropMap #: The property mapping
+    mapPair = Maps.ExpMap
```
Comment (lheagy) [inline SimPEG/EM/Base.py]: should the default be Maps.IdentityMap ?

File: SimPEG/EM/Base.py [hunk 5]
```diff
@@ -129,83 +172,83 @@ def MeMuI(self):
...
         """
         if getattr(self, '_MeMuI', None) is None:
-            self._MeMuI = self.mesh.getEdgeInnerProduct(self.curModel.mu, invMat=True)
+            self._MeMuI = self.mesh.getEdgeInnerProduct(self.mu, invMat=True)
         return self._MeMuI
 
-
     # ----- Electrical Conductivity ----- #
-    #TODO: hardcoded to sigma as the model
+    # TODO: hardcoded to sigma as the model
```
Comment (lheagy) [inline SimPEG/EM/Base.py]: we can remove this now?

File: SimPEG/EM/Base.py (hunk 4 above)
Comment (rowanc1) [inline SimPEG/EM/Base.py]: yeah.

Comment (rowanc1) [inline SimPEG/EM/Base.py]: thanks for catching that!

File: SimPEG/EM/Base.py (hunk 5 above)
Comment (rowanc1) [inline SimPEG/EM/Base.py]: Well, I suppose, sigma is still the only possible model, but it is a bit more explicit now!

File: SimPEG/EM/FDEM/ProblemFDEM.py [hunk 6]
```diff
@@ -162,24 +164,24 @@ def getSourceTerm(self, freq):
...
+            s_e[:, i] = s_e[:, i] + sei
 
         return s_m, s_e
 
 
-##########################################################################################
-################################ E-B Formulation #########################################
-##########################################################################################
+###############################################################################
+#                               E-B Formulation                               #
```
Comment (lheagy) [inline SimPEG/EM/FDEM/ProblemFDEM.py]: do we want to make this standard? perhaps start a wiki / contributors guide?

File: SimPEG/EM/Static/SIP/ProblemSIP.py [hunk 7]
```diff
@@ -1,59 +1,68 @@
...
-class ColeColePropMap(Maps.PropMap):
-    """
-        Property Map for EM Problems. The electrical conductivity (\\(\\sigma\\)) is the default inversion property, and the default value of the magnetic permeability is that of free space (\\(\\mu = 4\\pi\\times 10^{-7} \\) H/m)
-    """
 
-    eta = Maps.Property("Electrical Conductivity", defaultInvProp=True)
-    tau = Maps.Property("Electrical Conductivity", defaultVal=0.1, propertyLink=('taui', Maps.ReciprocalMap))
-    taui   = Maps.Property("Electrical Conductivity", defaultVal=1., propertyLink=('tau', Maps.ReciprocalMap))
-    c   = Maps.Property("Electrical Conductivity", defaultVal=1.)
+class BaseSIPProblem(BaseEMProblem):
```
Comment (lheagy) [inline SimPEG/EM/Static/SIP/ProblemSIP.py]: do we want to reference ColeCole specifically in the problem name? This implementation looks specific to the ColeCole parametrization of conductivity

File: SimPEG/Directives.py (hunk 1 above)
Comment (rowanc1) [inline SimPEG/Directives.py]: fixed in next commit.

File: SimPEG/EM/Base.py (hunk 3 above)
Comment (rowanc1) [inline SimPEG/EM/Base.py]: fixed.

File: SimPEG/EM/Static/SIP/ProblemSIP.py (hunk 7 above)
Comment (rowanc1) [inline SimPEG/EM/Static/SIP/ProblemSIP.py]: @sgkang??

File: SimPEG/FLOW/Richards/Empirical.py [hunk 8]
```diff
@@ -270,9 +290,16 @@ def setModel(self, m):
 
     def transform(self, u, m):
         self.setModel(m)
-        m = 1 - 1.0/self.n
-        f = ((  self.theta_s  -  self.theta_r  )/
-             ((1+abs(self.alpha*u)**self.n)**m)   +  self.theta_r)
+        m = 1 - 1.0 / self.n
+        f = (
```
Comment (lheagy) [inline SimPEG/FLOW/Richards/Empirical.py]: I find this hard to read mathematically - perhaps 
```
f = (
    (self.theta_s - self.theta_r) /
    ((1 + abs(self.alpha * u)**self.n)**m) + 
    self.theta_r
)
```

**Comment by rowanc1** [comment]
Yay, 80% test coverage!

**Comment by lheagy** [comment]
can we throw an error if you try to set curModel? @rowanc1

**Pull request merged**